konstraints-core/src/main/kotlin/tools/aqua/konstraints/smt/Context.kt

```diff
@@ -0,0 +1,768 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright 2023-2025 The Konstraints Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tools.aqua.konstraints.smt
+
+import tools.aqua.konstraints.parser.Theory
+import tools.aqua.konstraints.theories.BoolSort
+import tools.aqua.konstraints.util.Stack
+import tools.aqua.konstraints.util.zipWithSameLength
+
+private class CurrentContext {
+  val functions = mutableMapOf<String, SMTFunction<*>>()
+  val sorts = mutableMapOf<String, Sort>()
+}
+
+class Context {
+  private val forbiddenNames = mutableSetOf<String>()
+  private val currentContext = CurrentContext()
+  private val shadowingMap = Stack<MutableMap<SMTFunction<*>, SMTFunction<*>>>()
+  private val undoStack = Stack<MutableSet<SMTFunction<*>>>()
+  private var logic: Logic? = null
+
+  // true if we are currently in any binder (let/exists/forall/par/match)
+  private var activeBinderState = false
+
+  fun <T : Sort> addFunOrNull(func: SMTFunction<T>): SMTFunction<T>? {
+    return try {
+      addFun(func)
+    } catch (_: IllegalArgumentException) {
+      return null
+    } catch (_: IllegalStateException) {
+      return null
+    }
+  }
+
+  fun <T : Sort> addFun(func: SMTFunction<T>): SMTFunction<T> {
+    require(func.name !in forbiddenNames) {
+      "Can not overload or shadow theory symbol ${func.name}!"
+    }
+    require(func.name !in currentContext.functions) {
+      "Can not overload or shadow user defined symbol ${func.name}!"
+    }
+    check(!activeBinderState) { "Can not add functions to the current context in this state!" }
+
+    currentContext.functions[func.name] = func
+
+    // if the undoStack is not empty, and we are not currently inside any binder
+    // there was at least one push, so we save the added func to remove on the appropriate pop
+    if (undoStack.isNotEmpty()) {
+      undoStack.peek().add(func)
+    }
+
+    return func
+  }
+
+  fun contains(func: String) = currentContext.functions[func.trim('|')] != null
+
+  fun contains(expression: Expression<*>) = expression.func in currentContext.functions.values
+
+  fun contains(sort: Sort): Boolean = sort in currentContext.sorts.values
+
+  fun <T : Sort> getFuncOrNull(name: String, sort: T) =
+      try {
+        getFunc(name, sort)
+      } catch (_: FunctionNotFoundException) {
+        null
+      }
+
+  fun <T : Sort> getFunc(name: String, sort: T) =
+      currentContext.functions[name.trim('|')]?.castTo(sort)
```

Comment:
the context repeatedly handles SMT quoting (in a not fully correct manner); this should go elsewhere